### PR TITLE
adding option to use the bpe tokenizer tiktoken as mentioned in the lecture

### PR DESCRIPTION
**Pull request by jhlimm8**
tested it on down bad hyperparameters due to memory limitations and worked, the loss and number of parameters for both options are:

tiktoken:

0.452817 M parameters
step 0: train loss 10.8239, val loss 10.8242
step 500: train loss 8.5850, val loss 8.5881
step 1000: train loss 5.3027, val loss 5.3133
step 1500: train loss 3.6597, val loss 3.6863
step 2000: train loss 3.3955, val loss 3.4450
step 2500: train loss 3.3283, val loss 3.3641
step 3000: train loss 3.2395, val loss 3.2716
step 3500: train loss 3.1519, val loss 3.1619
step 4000: train loss 3.0349, val loss 3.0535
step 4500: train loss 2.9624, val loss 2.9659
step 4999: train loss 2.9048, val loss 2.9137

char:

0.001089 M parameters
step 0: train loss 4.1758, val loss 4.1754
step 500: train loss 3.4595, val loss 3.4740
step 1000: train loss 3.1641, val loss 3.1794
step 1500: train loss 2.9892, val loss 3.0016
step 2000: train loss 2.8817, val loss 2.8845
step 2500: train loss 2.8174, val loss 2.8264
step 3000: train loss 2.7864, val loss 2.7788
step 3500: train loss 2.7503, val loss 2.7482
step 4000: train loss 2.7327, val loss 2.7282
step 4500: train loss 2.7174, val loss 2.7107
step 4999: train loss 2.7026, val loss 2.6954

considering the magnitudes worth of increase in number of parameter for no final decrease in loss, its debatable whether its worthwhile to use it on low hyperparam settings at all, but could be worth for higher hyperparam settings as gpt2 and google(sentencepiece) uses it as mentioned in the lecture, or maybe its worth just for the sake of satisfying morbid curiosity

### CODE REVIEW

**Comment by ivofernandes** [comment]
So cool :)

Maybe the reason why the training is better on chars is because it has about 70 possible chars? And the tokenizer looks that generates about 50000 tokens, at least was what it looks like in my debugger
Screenshot 2023-03-05 at 18 32 25

And more options that makes the probabilistic chances of getting it right are just harder, but what about the generated text? maybe is about the same in the end of the training?

Or the maybe the model needs to be deeper if is dealing with tokens?

**Comment by Andrei-Aksionov** [comment]
As I see it the size of the dataset might be too small to give enough examples of each (or at least most) of the tokens for the model to learn. With a simple char tokenizer there are a lot of examples per each letter.
So, if you want to experiment, try to train tokenizer from scratch. [Here](https://discuss.huggingface.co/t/training-sentencepiece-from-scratch/3477) you can find an example. You might want to focus on the size of vocabulary.
